File: ios/Classes/scan/ScanView.m

#import "ScanView.h"
#import "ScanUtils.h"
@interface ScanView()<AVCaptureMetadataOutputObjectsDelegate>

@property(nonatomic , strong)AVCaptureSession * session;
@property(nonatomic , strong)FlutterMethodChannel * _channel;
@property(nonatomic , strong)ScanViewEventChannel * _event;
@property(nonatomic , strong)AVCaptureVideoPreviewLayer * captureLayer;
@property(nonatomic , strong)AVCaptureDevice * _device;

- (void)onMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result;

@end

@implementation ScanView

- (AVCaptureSession *)session{
    if(!_session){
        _session=[[AVCaptureSession alloc]init];
    }
    return _session;
}

- (instancetype)initWithFrame:(CGRect)frame viewIdentifier:(int64_t)viewId arguments:(id)args binaryMessenger:(NSObject<FlutterBinaryMessenger> *)messenger{
    if(self = [super initWithFrame:frame]){
        
       NSString * channelName=[NSString stringWithFormat:@"scanView/%lld/method",viewId];
        self._channel=[FlutterMethodChannel methodChannelWithName:channelName binaryMessenger:messenger];
        __weak __typeof__(self) weakSelf = self;
        [weakSelf._channel setMethodCallHandler:^(FlutterMethodCall * _Nonnull call, FlutterResult  _Nonnull result) {
            [weakSelf onMethodCall:call result:result];
        }];
        
        NSString * eventChannelName=[NSString stringWithFormat:@"scanView/%lld/event",viewId];
        FlutterEventChannel * _evenChannel = [FlutterEventChannel eventChannelWithName:eventChannelName binaryMessenger:messenger];
        self._event=[ScanViewEventChannel new];
        [_evenChannel setStreamHandler:self._event];
        
        AVCaptureVideoPreviewLayer * layer=[AVCaptureVideoPreviewLayer layerWithSession:self.session];
        self.captureLayer=layer;
        
        layer.backgroundColor=[UIColor blackColor].CGColor;
        [self.layer addSublayer:layer];
        layer.videoGravity=AVLayerVideoGravityResizeAspectFill;
        
        self._device=[AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
        AVCaptureDeviceInput * input=[[AVCaptureDeviceInput alloc] initWithDevice:self._device error:nil];
        AVCaptureMetadataOutput * output=[[AVCaptureMetadataOutput alloc]init];
        [self.session addInput:input];
        [self.session addOutput:output];
        self.session.sessionPreset=AVCaptureSessionPresetHigh;
        
        output.metadataObjectTypes=output.availableMetadataObjectTypes;
        [output setMetadataObjectsDelegate:self queue:dispatch_get_main_queue()];
        [output setMetadataObjectTypes:@[AVMetadataObjectTypeAztecCode,AVMetadataObjectTypeCode39Code,
                                         AVMetadataObjectTypeCode93Code,AVMetadataObjectTypeCode128Code, AVMetadataObjectTypeDataMatrixCode,AVMetadataObjectTypeEAN8Code,AVMetadataObjectTypeEAN13Code,AVMetadataObjectTypeITF14Code,AVMetadataObjectTypePDF417Code,AVMetadataObjectTypeQRCode,AVMetadataObjectTypeUPCECode]];
        
        [self.session startRunning];
    }
    return self;
}

- (void)layoutSubviews{
    [super layoutSubviews];
    self.captureLayer.frame=self.bounds;
}

-(void)onMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([call.method isEqualToString:@"startScan"]) {
        [self resume];
        result(nil);
    }else if([call.method isEqualToString:@"stopScan"]){
        [self pause];
        result(nil);
    }else if ([call.method isEqualToString:@"setFlashMode"]){
        NSNumber * status = [call.arguments valueForKey:@"status"];
        result([NSNumber numberWithBool:[self setFlashMode:[status boolValue]]]);
    }else if ([call.method isEqualToString:@"getFlashMode"]){
        result([NSNumber numberWithBool:[self getFlashMode]]);
    }else {
        result(FlutterMethodNotImplemented);
    }
}

-(void)resume{
    if(![self.session isRunning]){
        [self.session startRunning];
    }
}

-(void)pause{
    if ([self.session isRunning]) {
        [self.session stopRunning];
    }
}

-(BOOL)setFlashMode:(BOOL) status{
    [self._device lockForConfiguration:nil];
    BOOL isSuccess = YES;
    if ([self._device hasFlash]) {
        if (status) {
            self._device.flashMode=AVCaptureFlashModeOn;
            self._device.torchMode=AVCaptureTorchModeOn;
        }else{
            self._device.flashMode = AVCaptureFlashModeOff;
            self._device.torchMode = AVCaptureTorchModeOff;
        }
    }else{
        isSuccess=NO;
    }
    [self._device unlockForConfiguration];
    
    return isSuccess;
    
}
-(BOOL)getFlashMode{
    [self._device lockForConfiguration:nil];
    BOOL isSuccess = self._device.flashMode==AVCaptureFlashModeOn&&
    self._device.torchMode==AVCaptureTorchModeOn;
    [self._device unlockForConfiguration];
    return isSuccess;
    
}
- (void)captureOutput:(AVCaptureOutput *)output didOutputMetadataObjects:(NSArray<__kindof AVMetadataObject *> *)metadataObjects fromConnection:(AVCaptureConnection *)connection{
    if (metadataObjects.count>0) {
        AVMetadataMachineReadableCodeObject * data=metadataObjects[0];
        NSString * value=data.stringValue;
        if(value.length&&self._event){
            [self._event getResult:[ScanUtils scanDataToMap:data]];
        }
    }
}
@end



@implementation ScanViewEventChannel

- (FlutterError *)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)events{
    self.events = events;
    if(self.scanView){
        NSNumber * isScan=[arguments valueForKey:@"isScan"];
        if(isScan){
            if (isScan.boolValue) {
                [self.scanView resume];
            }else{
                [self.scanView pause];
            }
        }
    }
    return nil;
}

- (FlutterError * _Nullable)onCancelWithArguments:(id _Nullable)arguments {
    if(self.scanView){
        [self.scanView pause];
    }
    return nil;
}

- (void)getResult:(NSDictionary *)msg{
    if(self.events){
        self.events(msg);
    }
}


@end

File: ios/Classes/scan/ScanView.swift
import AVFoundation
import Flutter


class ScanView: UIView,AVCaptureMetadataOutputObjectsDelegate {
    private let scanEvent = ScanEventStreamHandler()
    private var session: AVCaptureSession
    private var methodChannel: FlutterMethodChannel
    private var captureLayer: AVCaptureVideoPreviewLayer
    private var device: AVCaptureDevice
    
    init(frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?, binaryMessenger: (NSObjectProtocol & FlutterBinaryMessenger)?) {
        let methodchannelName = String(format: "scanView/%lld/method", viewId)
        let eventChannelName = String(format: "scanView/%lld/event", viewId)
        let eventChannel = FlutterEventChannel(name: eventChannelName, binaryMessenger:binaryMessenger!)
        methodChannel = FlutterMethodChannel(name: methodchannelName, binaryMessenger: binaryMessenger!)
        eventChannel.setStreamHandler((scanEvent as! FlutterStreamHandler & NSObjectProtocol))
        
//        weak var weakSelf = self
//        weakSelf.channel.methodCallHandler = { call, result in
//            weakSelf.on(call, result: result)
//        }

        captureLayer = AVCaptureVideoPreviewLayer(session: session)
        captureLayer.backgroundColor = UIColor.black.cgColor
        captureLayer.addSublayer(captureLayer)
        captureLayer.videoGravity = .resizeAspectFill
        captureLayer.backgroundColor = UIColor.black.cgColor
        device = AVCaptureDevice.default(for: .video)!
        var input: AVCaptureDeviceInput? = nil
        do {
            input = try AVCaptureDeviceInput(device: device)
        } catch {
        }
        let output = AVCaptureMetadataOutput()
        if let input = input {
            session.addInput(input)
        }
        session.addOutput(output)
        session.sessionPreset = .high
        output.metadataObjectTypes = output.availableMetadataObjectTypes
        output.setMetadataObjectsDelegate(self, queue: DispatchQueue.main)
        output.metadataObjectTypes = [.aztec, .code39,.code93, .code128,.dataMatrix,.ean8,.ean13,.itf14,.pdf417,.qr,.upce]
        startScan()
//        return self
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        captureLayer.frame = bounds
    }
    
    
     func handle( call: FlutterMethodCall, result: FlutterResult) {
        if (call.method == "startScan") {
            startScan()
            result(nil)
        } else if (call.method == "stopScan") {
            pause()
            result(nil)
        } else if (call.method == "setFlashMode") {
            let status = (call.arguments as AnyObject).value(forKey: "status") as? NSNumber
            result(NSNumber(value: status?.boolValue ?? false))
        } else if (call.method == "getFlashMode") {
            result(NSNumber(value: getFlashMode()))
        } else {
            result(FlutterMethodNotImplemented)
        }
    }
    
    func startScan() {
        if !session.isRunning {
            session.startRunning()
        }
    }
    
    func stopScan() {
        if session.isRunning {
            session.stopRunning()
        }
    }
    
    func setFlashMode( status: Bool) -> Bool {
        do {
            try device.lockForConfiguration()
        } catch {
        }
        var isSuccess = true
        if device.hasFlash {
            if status {
                device.flashMode = .on
                device.torchMode = .on
            } else {
                device.flashMode = .off
                device.torchMode = .off
            }
        } else {
            isSuccess = false
        }
        device.unlockForConfiguration()
        
        return isSuccess
        
    }
    
    func getFlashMode() -> Bool {
        do {
            try device.lockForConfiguration()
        } catch {
        }
        let isSuccess = device.flashMode == .on && device.torchMode == .on
        device.unlockForConfiguration()
        return isSuccess
        
    }
    func captureOutput(output: AVCaptureMetadataOutput, didOutput metadataObjects: [AVMetadataObject], from connection: AVCaptureConnection) {
        if metadataObjects.count > 0 {
            let data = metadataObjects[0] as? AVMetadataMachineReadableCodeObject
            let value = data?.stringValue
            if (value?.count ?? 0) != 0  {
                scanEvent.sendEvent(event: ScanUtils.scanDataToMap(data))
            
            }
        }
    }
    //    func onCancelWithArguments(arguments: Any?) -> FlutterError? {
    //        if scanView {
    //            scanView.pause()
    //        }
    //        return nil
    //    }
    //
    //    func getResult(msg: [AnyHashable : Any]?) {
    //        if events {
    //            events(msg)
    //        }
    //    }
    
}

import Foundation
class ScanEventStreamHandler: FlutterStreamHandler {
    private var eventSink:FlutterEventSink? = nil
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
    
    //发送event
    public func sendEvent(event:Any) {
        eventSink?(event)
    }
}


//class ScanViewEventChannel : NSObject<FlutterStreamHandler> {
//    private var eventSink: FlutterEventSink
//
//
//    override func onListenWithArguments( arguments: Any, events: FlutterEventSink) -> FlutterError? {
//        eventSink = events
//        if ScanView {
//            let isScan = (arguments as? NSObject)?.value(forKey: "isScan") as? NSNumber
//            if isScan != nil {
//                if isScan?.boolValue ?? false {
//                    ScanView.resume()
//                } else {
//                    ScanView.pause()
//                }
//            }
//        }
//        return nil
//    }
//
//    func onCancelWithArguments(arguments: Any) -> FlutterError? {
//        if ScanView {
//            ScanView.pause()
//        }
//        return nil
//    }
//
//    func getResult(msg: [AnyHashable : Any]) {
//        if eventSink {
//            eventSink(msg)
//        }
//    }
//}
